File: framework/test2/char/chars2d/Chars2d.java
import soda.unittest.*;

import java.util.*;
import java.util.function.Supplier;
import java.util.stream.*;

import soda.leetcode.*;
import soda.unittest.Validators;
import soda.unittest.work.TestWork;

import static soda.unittest.LoggerHelper.logger;

class Solution {
    public char[][] toUpper(List<List<Character>> matrix) {
        int diff = 'a' - 'A';
        char[][] mx = new char[matrix.size()][matrix.get(0).size()];
        for (int i = 0; i < matrix.size(); ++i) {
            for (int j = 0; j < matrix.get(0).size(); ++j) {
                mx[i][j] = (char) (matrix.get(i).get(j) - diff);
            }
        }
        return mx;
    }
}

public class Chars2d implements Supplier<TestWork> {

    @Override
    public TestWork get() {
        var work = new TestWork(new Solution(), "toUpper");
        // var work = TestWork.forStruct(Struct.class);
        // work.setValidator((e, r) -> {...});
        work.setCompareSerial(true);
        return work;
    }

    public static void main(String[] args) throws Exception {
        new Chars2d().get().run();
    }

}

File: framework/test2/char/chars2d/Chars2d.scala
import soda.scala.unittest._

import scala.collection.mutable
import scala.reflect.runtime.universe.typeOf

class Solution {}
object Solution {
  def toUpper(matrix: List[List[Char]]): Array[Array[Char]] = {
    val diff = 'a' - 'A'
    val mx = Array.ofDim[Char](matrix.size, matrix(0).size)
    for (i <- matrix.indices) {
      for (j <- matrix(0).indices) {
        mx(i)(j) = (matrix(i)(j) - diff).toChar
      }
    }
    return mx
  }
}

class Chars2d {
  def get(): TestWork = {
    val work = new TestWork(typeOf[Solution], "toUpper")
    // val work = TestWork.forStruct(...)
    // work.setValidator((R, R) => Boolean)
    work.compareSerial = true
    work
  }
}

object Chars2d extends App {
  new Chars2d().get().run()
}
